### DISCUSSION TITLE
Unable to ping dockerized vpn client from other host vpn client (other computer - non dockerized vpn client)

### DISCUSSION THREAD

**Discussion by jasek5** [qdm12/gluetun#626] [Q&A] [answered]
![current-situation](https://user-images.githubusercontent.com/12669491/133328076-279e668d-2442-4700-a242-1471f4ea4c6b.jpg)

Hi - I have dockerized openvpn server (https://github.com/kylemanna/docker-openvpn) on vps 1 where I generated 3 different certificates for 3 different clients (notebook, vps2(dockerized client) and vps3) - each of those have static vpn ip done by these commands (https://github.com/kylemanna/docker-openvpn/blob/master/docs/static-ips.md).

Each client has correct vpn ip and can connect to the internet, but the problem is that the dockerized openvpn client is unpingable from other vpn clients, but dockerized openvpn client can ping other openvpn clients via their openvpn ip.(For better understanding I created simple diagram available here on imgur:[diagram with logs - blue is docker container and orange is host](https://imgur.com/0XNP8G5).

My **goal** is that I would like to make each client pingable from each other openvpn client by its vpn ip especially the containerized one, but only non-dockerized openvpn clients can ping each other without problem.

If you need any other outputs from commands please let me know and I will provide those for you. Thank you very much for your time.

My openvpn server config:

```
server 192.168.255.0 255.255.255.0

verb 3

key /etc/openvpn/pki/private/example.key

ca /etc/openvpn/pki/ca.crt

cert /etc/openvpn/pki/issued/example.crt

dh /etc/openvpn/pki/dh.pem

tls-auth /etc/openvpn/pki/ta.key

key-direction 0

keepalive 10 60

persist-key

persist-tun



proto udp

# Rely on Docker to do port mapping, internally always 1194

port 1194

dev tun0

status /tmp/openvpn-status.log



user nobody

group nogroup

comp-lzo no



### Route Configurations Below

route 192.168.254.0 255.255.255.0



### Push Configurations Below

push "block-outside-dns"

push "dhcp-option DNS 8.8.8.8"

push "dhcp-option DNS 8.8.4.4"

push "comp-lzo no"



### Extra Configurations Below

client-to-client
```

And this is my gluetun openvpn docker-compose file:version: "3.7"

```
services:

gluetun:

image: qmcgaw/gluetun

container_name: gluetun

devices:

- /dev/net/tun

cap_add:

- NET_ADMIN

network_mode: bridge

volumes:

- /var/infrastructure/docker/volumes/gluetun:/gluetun

environment:

- VPNSP=private internet access

- VPN_TYPE=openvpn

- OPENVPN_ROOT=yes

- OPENVPN_INTERFACE=tun1

- OPENVPN_CUSTOM_CONFIG=/gluetun/client2.conf

# Timezone for accurate logs times

- "TZ=Europe/Prague"

restart: always
```

I guess that it will be probably a routing problem, but i am quite a newbie in this area...

I am attaching ip route output here and ip information from the diagram above - just in case.

## VPS1 - host of dockerized openvpn server

ip of host where is dockerized opnvpn SERVER is:

`89.221.215.68/24`

OUTPUT of ip route :

```
default via 89.221.215.1 dev eth0 onlink

89.221.215.0/24 dev eth0 proto kernel scope link src 89.221.215.68

172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown

172.19.0.0/16 dev br-af92657b125f proto kernel scope link src 172.19.0.1

172.21.0.0/16 dev br-e85b4bce6f68 proto kernel scope link src 172.21.0.1

172.22.0.0/16 dev br-d25bbe41de5e proto kernel scope link src 172.22.0.1

172.23.0.0/16 dev br-66cfa683dc5e proto kernel scope link src 172.23.0.1

172.26.0.0/16 dev br-315f338b4741 proto kernel scope link src 172.26.0.1 linkdown

172.27.0.0/16 dev br-c7c158298da1 proto kernel scope link src 172.27.0.1 linkdown

172.172.0.0/16 dev br-3d9dcfa49be3 proto kernel scope link src 172.172.0.1

192.168.240.0/20 dev br-04b651f8d304 proto kernel scope link src 192.168.240.1
```

------------

ip of docker container is `172.172.0.15/16 (eth1)` and `172.23.0.2/16 (eth0)`

ip of vpn ip - server

```
tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN group default qlen 100

link/none

inet 192.168.255.1 peer 192.168.255.2/32 scope global tun0

valid_lft forever preferred_lft forever
```

OUTPUT of ip route :

```
default via 172.23.0.1 dev eth0

172.23.0.0/16 dev eth0 proto kernel scope link src 172.23.0.2

172.172.0.0/16 dev eth1 proto kernel scope link src 172.172.0.15

192.168.254.0/24 via 192.168.255.2 dev tun0

192.168.255.0/24 via 192.168.255.2 dev tun0

192.168.255.2 dev tun0 proto kernel scope link src 192.168.255.1
```

-----------

## VPS2

ip of host where is dockerized openvpn CLIENT is:

`172.16.44.101 `(ens18 device)

OUTPUT of ip route :

```
default via 172.16.44.1 dev ens18 proto dhcp src 172.16.44.101 metric 100

172.16.44.0/24 dev ens18 proto kernel scope link src 172.16.44.101

172.16.44.1 dev ens18 proto dhcp scope link src 172.16.44.101 metric 100

172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1

172.18.0.0/16 dev br-736451a455a5 proto kernel scope link src 172.18.0.1

172.19.0.0/16 dev br-7721fec72e4e proto kernel scope link src 172.19.0.1

172.20.0.0/16 dev br-9681670259ca proto kernel scope link src 172.20.0.1

172.21.0.0/16 dev br-f5c0b6ac86b4 proto kernel scope link src 172.21.0.1

172.22.0.0/16 dev br-5e55ad1f9e41 proto kernel scope link src 172.22.0.1 linkdown

172.23.0.0/16 dev br-d19acaaab5e9 proto kernel scope link src 172.23.0.1

172.24.0.0/16 dev br-3116d40bc24e proto kernel scope link src 172.24.0.1 linkdown

172.25.0.0/16 dev br-14c4f737eb9f proto kernel scope link src 172.25.0.1

172.26.0.0/16 dev br-bed87c01172f proto kernel scope link src 172.26.0.1 linkdown

172.27.0.0/16 dev br-80ae56cf9900 proto kernel scope link src 172.27.0.1

172.28.0.0/16 dev br-6c7198b680d7 proto kernel scope link src 172.28.5.254

172.30.0.0/16 dev br-831554646ccf proto kernel scope link src 172.30.0.1 linkdown

192.168.48.0/20 dev br-836b516f7988 proto kernel scope link src 192.168.48.1

192.168.64.0/20 dev br-f4893b5ee801 proto kernel scope link src 192.168.64.1 linkdown

192.168.144.0/20 dev br-2a6b74734bb4 proto kernel scope link src 192.168.144.1 linkdown
```

------------

ip of docker container is `172.17.0.3/16`

vpn ip of container info output:
```

tun1: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 500

link/none

inet 192.168.254.5 peer 192.168.254.6/32 scope global tun1

valid_lft forever preferred_lft forever
```

ip route inside dockerized openvpn client:

```
0.0.0.0/1 via 192.168.254.6 dev tun1

default via 172.17.0.1 dev eth0

89.221.215.68 via 172.17.0.1 dev eth0

128.0.0.0/1 via 192.168.254.6 dev tun1

172.17.0.0/16 dev eth0 proto kernel scope link src 172.17.0.2

192.168.254.0/24 via 192.168.254.6 dev tun1

192.168.254.6 dev tun1 proto kernel scope link src 192.168.254.5

192.168.255.0/24 via 192.168.254.6 dev tun1

```
------------

## VPS3

ip of host where is installed openvpn client without docker is

`172.16.44.102/24`

vpn ip info output:

```
tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 100

link/none

inet 192.168.254.13 peer 192.168.254.14/32 scope global tun0

valid_lft forever preferred_lft forever

inet6 fe80::9dc1:737b:f9e7:2584/64 scope link stable-privacy

valid_lft forever preferred_lft forever

```

ip route of vps3:

```
0.0.0.0/1 via 192.168.254.14 dev tun0

default via 172.16.44.1 dev ens18 proto dhcp src 172.16.44.102 metric 100

89.221.215.68 via 172.16.44.1 dev ens18

128.0.0.0/1 via 192.168.254.14 dev tun0

172.16.44.0/24 dev ens18 proto kernel scope link src 172.16.44.102

172.16.44.1 dev ens18 proto dhcp scope link src 172.16.44.102 metric 100

172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown

172.18.0.0/16 dev br-c39d54a7050a proto kernel scope link src 172.18.0.1 linkdown

172.19.0.0/16 dev my-bridge proto kernel scope link src 172.19.0.1 linkdown

192.168.254.0/24 via 192.168.254.14 dev tun0

192.168.254.14 dev tun0 proto kernel scope link src 192.168.254.13

192.168.255.0/24 via 192.168.254.14 dev tun0
```

-----------

## Notebook

ip of notebook is `192.168.83.180/24`

vpn ip info output:

```
tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 500

link/none

inet 192.168.254.1 peer 192.168.254.2/32 scope global tun0

valid_lft forever preferred_lft forever

inet6 fe80::22a4:5535:ab1b:bff9/64 scope link stable-privacy

valid_lft forever preferred_lft forever
```

OUTPUT of ip route :

```
0.0.0.0/1 via 192.168.254.2 dev tun0

default via 192.168.83.124 dev wlp0s20f3 proto dhcp metric 600

89.221.215.68 via 192.168.83.124 dev wlp0s20f3

128.0.0.0/1 via 192.168.254.2 dev tun0

192.168.83.0/24 dev wlp0s20f3 proto kernel scope link src 192.168.83.180 metric 600

192.168.254.0/24 via 192.168.254.2 dev tun0

192.168.254.2 dev tun0 proto kernel scope link src 192.168.254.1

192.168.255.0/24 via 192.168.254.2 dev tun0
```

----------

**Comment by qdm12** [comment, marked as answer]
Hi there! I haven't read through all your information (yet), but maybe try turning off the gluetun firewall with `FIREWALL=off`? Let me know if this allows it to receive pings. If not, it might be a routing problem perhaps.

**Reply by jasek5** [reply]
wow - `FIREWALL=off` did the trick - thank you very much !

**Reply by qdm12** [reply]
Wait a second 😄 Disabling the firewall means some traffic might leak out the vpn. If you don't care then all good. Otherwise, you can try `docker exec -it gluetun /bin/sh` and run some `iptables` commands to see what allows it to receive pings. If you're not familiar with it, let me know I'll write you some iptables commands to try.

**Reply by qdm12** [reply]
Maybe:

```sh
iptables -A INPUT -p icmp --icmp-type 8 -i tun0 -j ACCEPT
```

would make it work together with the firewall on. Command taken from https://www.cyberciti.biz/tips/linux-iptables-9-allow-icmp-ping.html

EDIT: let me know and I can add it as an optional env variable to allow receiving ping requests.

**Reply by jasek5** [reply]
Thank you for the concern - I am not familiar with iptables.
My usage is to make service (like ldap i.e. - not ldaps) available over openvpn to certain dockerized services running on another vps than the containerized ldap (i.e).

**Reply by jasek5** [reply]
I can confirm that the following command works with omitting `FIREWALL=off`
`iptables -A INPUT -p icmp --icmp-type 8 -i tun0 -j ACCEPT`

**Reply by qdm12** [reply]
Cool I'll add that command to the code, disabled by default (we don't like receiving pings by default). Is `FIREWALL_ALLOW_INCOMING_ICMP` ok? Feel free to suggest a better name.

**Reply by jasek5** [reply]
Well the name of the variable is fine - I was just thinking if there could be a way to run certain amount of commands (like more custom iptables rules) on each docker-compose up ..etc without manually editing the dockerfile and building the own image (of course based on your gluetun image), but maybe this is not exactly dockerway or I am maybe missing some know-how about docker (more likely).

EDIT: because i.e I can ping the gluetun client, but a I cannot connet to ldap over `<vpn-client-ip>:389` - which I guess means that I need more ip tables rules (however I have not tested the ldap connection with `FIREWALL=off`

**Reply by qdm12** [reply]
Actually you can already do that: https://github.com/qdm12/gluetun/wiki/Advanced-setup#custom-iptables-rules

I think this might be best left like this, since your use case is a bit singular (and also I'm lazy haha).

**Reply by jasek5** [reply]
I understand that :D sorry I somehow missed that section... Thank you very much again.
You have done enough already :) 

EDIT: so I only need to bindmount the file `/iptables/post-rules.txt` - is that correct ?

**Reply by qdm12** [reply]
Yep that's it. I reworked the entire wiki today (the page moved to https://github.com/qdm12/gluetun/wiki/Firewall-options#custom-iptables-rules) since it was confusing to browse with the amount of dust it took in the last few months 😄 Hopefully it's clearer now!

**Reply by Cknight70** [reply]
@qdm12 The responses here were very helpful for me, I'm wondering did you ever make FIREWALL_ALLOW_INCOMING_ICMP a variable? or should I do custom ip tables?
